publish.js

```diff
@@ -1,96 +1,99 @@
-    const spec = npa.resolve(manifest.name, manifest.version)
-    // NOTE: spec is used to pick the appropriate registry/auth combo.
-    opts = opts.concat(manifest.publishConfig, { spec })
+async function publish (folder, manifest, opts) {
+  if (manifest.private) {
+    throw Object.assign(
+      new Error(
+        `This package has been marked as private\n
+         Remove the 'private' field from the package.json to publish it.`
+      ),
+      { code: 'EPRIVATE' }
+    )
+  }
+
+  // spec is used to pick the appropriate registry/auth combo
+  const spec = npa.resolve(manifest.name, manifest.version)
+  opts = {
+    defaultTag: 'latest',
+    // if scoped, restricted by default
+    access: spec.scope ? 'restricted' : 'public',
+    algorithms: ['sha512'],
+    ...opts,
+    spec
+  }
+
+  const stat = await statAsync(folder)
+
+  // checks if it's a dir
+  if (stat.isDirectory()) {
     const reg = npmFetch.pickRegistry(spec, opts)
     const auth = npmAuth(reg, opts)
-    const pubManifest = patchedManifest(spec, auth, manifest, opts)
+    const pubManifest = patchManifest(manifest, opts)
 
-    // registry-frontdoor cares about the access level, which is only
-    // configurable for scoped packages
+    // registry-frontdoor cares about the access level,
+    // which is only configurable for scoped packages
     if (!spec.scope && opts.access === 'restricted') {
       throw Object.assign(
         new Error("Can't restrict access to unscoped packages."),
         { code: 'EUNSCOPED' }
       )
     }
 
-    return slurpTarball(tarball, opts).then(tardata => {
-      const metadata = buildMetadata(
-        spec, auth, reg, pubManifest, tardata, opts
-      )
-      return npmFetch(spec.escapedName, opts.concat({
+    const tarballData = await pacote.tarball(`file:${folder}`)
+    const metadata = buildMetadata(
+      auth, reg, pubManifest, tarballData, opts
+    )
+
+    try {
+      return await npmFetch(spec.escapedName, {
+        ...opts,
         method: 'PUT',
         body: metadata,
         ignoreBody: true
-      })).catch(err => {
-        if (err.code !== 'E409') { throw err }
-        return npmFetch.json(spec.escapedName, opts.concat({
-          query: { write: true }
-        })).then(
-          current => patchMetadata(current, metadata, opts)
-        ).then(newMetadata => {
-          return npmFetch(spec.escapedName, opts.concat({
-            method: 'PUT',
-            body: newMetadata,
-            ignoreBody: true
-          }))
-        })
       })
-    })
-  }).then(() => true)
+    } catch (err) {
+      if (err.code !== 'E409') { throw err }
+      const current = await npmFetch.json(spec.escapedName, {
+        ...opts,
+        query: { write: true }
+      })
+      const newMetadata = await patchMetadata(current, metadata, opts)
+      return npmFetch(spec.escapedName, {
+        ...opts,
+        method: 'PUT',
+        body: newMetadata,
+        ignoreBody: true
+      })
+    }
+  } else {
```

Comment:
It'd be a little clearer to put this else block and the throw up above the "if it's a dir" section.

So, instead of this:

```js
if (it's a dir) {
  long
  bunch
  of
  stuff
  .....
} else {
  throw 'not a dir!'
}
```

it'd be this:

```js
if (not a dir) {
  throw 'not a dir!'
}

// now the rest of the stuff
```

so that the reader doesn't have to hold that context in their head the whole time.